### PR TITLE
Reader: new ui reducer for expanded states on cards

### PR DESCRIPTION
**Pull request by samouri**
This PR aims to add the expanded state of cards into the redux store.
That way the expanded state isn't prone to being lost if a component gets switched out during a re-render.

TODO:
 - [x] un-expand on stream load
 - [x] add unit tests for reducer
 - ~~[ ] Make work with videos cards (`showEmbed` --> `isExpanded`)~~
 - ~~[ ] verify all the stats are working as expected. Add `calypso_reader_video_expanded`~~

Since this is a bug currently live in production it would definitely be better to address the second two in a followup PR

fix for https://github.com/Automattic/wp-calypso/issues/12333.

### CODE REVIEW

**Comment by samouri** [comment]
Overview
------
1. new reducer that basically acts as a set of currently expanded cards in the current stream.  Its at `ui.reader.cardExpansions`
2. new selector for `isReaderCardExpanded( postKey )`
3. `connect` the reader post card so that it can know if its expanded / dispatch an expansion
4. `connect` Streams so that they can reset the expanded state when they mount (every time stream loads should be a blank slate)

File: client/state/ui/reader/card-expansions/test/reducer.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		const nextState = reducer( prevState, action );
+		expect( nextState ).eql( {} );
+	} );
+
+	it( 'should add a newly expanded card to state', () => {
+		const action = {
+			type: READER_EXPAND_CARD,
+			payload: { postKey }
+		};
+		const prevState = {};
```
Comment (blowery) [inline client/state/ui/reader/card-expansions/test/reducer.js]: deep freeze prior states

File: client/state/ui/reader/card-expansions/test/reducer.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		const nextState = reducer( prevState, action );
+		expect( nextState ).eql( {
+			[ keyToString( postKey ) ]: true,
+		} );
+	} );
+
+	it( 'should clear the entire state on a reset', () => {
+		const action = { type: READER_RESET_CARD_EXPANSIONS };
+		const prevState = {
```
Comment (blowery) [inline client/state/ui/reader/card-expansions/test/reducer.js]: freeze this one too

**Comment by samouri** [comment]
@blowery, all frozen.  Functionality looked good though?
specifically:
1. photo cards expand.  switching between streams after expanding cards behave as expected
2. photo card expansion metrics are still emitted

**Comment by blowery** [comment]
yup, everything looked good

**Pull request merged**